### PR TITLE
Big sweep: migrate codebase to Jakarta EE 10 and use Payara 6 now (third iteration)

### LINKED ISSUE #8305: Upgrade to Payara 6 and from EE 8 to EE 10 (end of Payara 5 Community Edition security patches Q2/2022)
As a "bklog: Deliverable" This is decomposed into smaller issues, grouped under [D: Payara 6 Upgrade](https://github.com/IQSS/dataverse/labels/D%3A%20Payara%206%20Upgrade)

---

Hat tip to @poikilotherm who [posted][] in chat a link to a video in which Steve Millidge, founder of Payara, answered a question about how long Payara 5 Community Edition (which we recommend in the guides) will receive security patches: https://youtu.be/juD0XNNq344?t=1972

[posted]: https://matrix.to/#/!AmypvmJtUjBesRrnLM:matrix.org/$YqXmiwT6hoyVLfyKuH7rQVKrOFFRHBGFoJqfXR-wjM0

I've transcribed the question and answer:

Q: "For the Community Edition, when will v5 stop getting security patches?"

A: "Payara 5 in Community was basically... when we switch the main branch over to 6, which will be on the point of 6 release, then Community will march forward on Payara 6. So Payara 5 patches, if you like, will stop at that point through Community. So then it will be Enterprise where you'll get patches and security fixes for Payara 5. Payara Community essentially is a marching version so whenever the latest release is the latest release."

I didn't watch the whole video but @poikilotherm says Payara plans to switch the main branch to 6 in Q2/2022.

(By the way, #8064 is the issue where we're tracking the next 5.x upgrade from Payara. As of this writing we are advising installations to use Payara 5.2021.5: https://guides.dataverse.org/en/5.9/installation/prerequisites.html#payara )

### LINKED ISSUE #9615: create clean Payara 6 branch
We've been discussing at tech hours and elsewhere that we don't have a Payara 6 branch that is clean from a git perspective. We should create one.

This is the most recent attempt at a Payara 6 branch:

- #9116

### LINKED ISSUE #9755: In Docker, can't run a harvesting job, com.sun.aas.instanceRoot
If you try to run the API test suite against Dataverse when running in Docker you will notice that HarvestingClientsIT#testHarvestingClientRun will fail.

The expected error in server.log (again, only when Dataverse is running in Docker) is something like this:

`dev_dataverse>   Caught exception running an asynchronous harvest (dataverse "h85b8d3b"): java.nio.file.NoSuchFileException|#]`

(If you edit the code to print out the exception you see something like this: `: ../logs/harvest_h85b8d3b_2023-08-01T18-33-45.log.lck`)

This is because HarvesterServiceBean.java uses `../log` to find `domain1/logs` which doesn't work in Docker. Using `System.getProperty("com.sun.aas.instanceRoot")` seems to fix it.

This has been discussed in Slack at https://iqss.slack.com/archives/C010LA04BCG/p1690912135247139 where @landreev and I decided to put the fix in the Paraya 6 PR: https://github.com/IQSS/dataverse/pull/9685/commits/8efb96487d181f83726ac5a338a13692b1710850

I'm going to mark that Payara 6 PR as closing this issue.

It's still a problem in develop/5.13/5.14 but since this only affects Dataverse running in Docker we can live with it.

Note that there are many places in the code where we use this `../log` trick or similar. We can open issue for these as we find them.

### PR DESCRIPTION
**Pull request by poikilotherm**
~~THIS IS READY BUT SHOULD NOT BE MARKED NON-DRAFT UNTIL 5.14 HAS BEEN RELEASED~~

**What this PR does / why we need it**:

Payara 5 is going EOL for community usage (= without a payed support contract). Payara 6 is the successor, but allows to use Jakarta EE 10 only. Issue #8305 is the deliverable, where this PR is a part of.

At tech hour we agreed that we want to do a clean cut after the 5.14 release and migrate develop to Jakarta EE 10 and hack on it further. This PR prepares this step.

This PR is a successor to PR #9116 

**Which issue(s) this PR closes**:

- Closes #8305
- Closes #9755

**Special notes for your reviewer**:

Note that as 5.14 is not yet released, this PR will need an update from `develop` once done and will need to add some more migrations of the outstanding code parts that are due for 5.14. (These are minor, so we started the branch)

Open Tasks this PR did not address (and maybe should be fed back to #8305):
- ~~Should we preserve backward compatibility of the SWORD API by inserting `attachment; ` for the user if it isn't supplied?~~ DONE, see https://github.com/IQSS/dataverse/pull/9685#issuecomment-1656340152
- ~~The Perl installer has not been updated. Honestly, @pdurbin and I are ok with deleting it (and references to it, of course, if any).~~ DONE, deleted. docker-aio and docker-dcm too.
- ~~The release note snippet needs work. Upgrades have not been tested.~~ DONE. Release note snippet added. There's a dedicated issue for upgrades: #9340

**Suggestions on how to test this**:

- API tests at https://jenkins.dataverse.org/job/IQSS-Dataverse-Payara6
- Will need a great review to make sure we touch every corner of the UI, as Jakarta Faces is notably more strict

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:

(No)

**Is there a release notes update needed for this change?**:

Yes, a stub is included, but needs work. Maybe can be done as part of another issue for Dataverse 6 release notes.

**Additional documentation**:

None but references in the guides to Payara 5 have been switched to Payara 6.

### CODE REVIEW

**Comment by poikilotherm** [comment]
Note this image will not work- we don't rebuild base image in the flow and we don't (yet?) have P6 images on the hub.

**Comment by poikilotherm** [comment]
I just updated:
- Merged latest `develop`
- Update to 6.2023.7

**Comment by pdurbin** [comment]
> Should we preserve backward compatibility of the SWORD API by inserting attachment; for the user if it isn't supplied?

Just a heads up that I reverted 89182c1 (in 83c55a9) and implemented Oliver's suggestion above in 0221006

API test are passing!

The release note has gone away as part of the revert so we should add it back in at some point.

Oh, before that I merged from develop and in da5193f added a fix for this issue:

- https://github.com/payara/Payara/issues/6344

**Comment by pdurbin** [comment]
> The Perl installer has not been updated. Honestly, @pdurbin and I are ok with deleting it (and references to it, of course, if any).

Here's a preview of what I'm proposing that we remove: https://github.com/IQSS/dataverse/compare/8305-payara6-ee10-v3...8305-rm-perl-docker-aio

- Perl installer
- docker-aio (relies on the Perl installer)
- docker-dcm (relies on docker-aio)

We want to deprecate Rsync/DCM/RSAL anyway:

- #8985

Slack discussion: https://iqss.slack.com/archives/C010LA04BCG/p1691009854451909

File: src/test/java/edu/harvard/iq/dataverse/api/UtilIT.java [hunk 1]
```diff
@@ -2577,9 +2582,24 @@ static boolean sleepForReexport(String idOrPersistentId, String apiToken, int du
...
+                i++;
+                if (i > duration) {
+                    break;
+                }
+            } catch (InterruptedException ex) {
+                Logger.getLogger(UtilIT.class.getName()).log(Level.SEVERE, null, ex);
+            }
+        } while (true);
+        return i <= duration;
+    }
```
Comment (pdurbin) [inline src/test/java/edu/harvard/iq/dataverse/api/UtilIT.java:2602]: This sleepForDeadlock is used in DataversesIT. Details:

commit 7b4f827af1a885a2560ba717b52d7d34fe03ab2c
Author: Philip Durbin <philip_durbin@harvard.edu>
Date:   Fri Jun 3 12:47:28 2022 -0400

    stop using `@Context HttpServletRequest`, add sleep #8305
    
    In EditDDI I was seeing a NullPointerException on line 136
    (createNewDraftVersion). Switch from `@Context HttpServletRequest`
    to our ordinary way of API auth fixed it.
    
    In DataversesIT I was seeing an OptimisticLockException so I
    added a sleep.

**Review by pdurbin — COMMENTED** [review]
Phew! For this review I simply scanned through the 639 (!) changed files looking for anything NOT related to the javax to jakarta namespace change. I left a lot of comments, some of which I'll probably follow up on with commits.

File: doc/sphinx-guides/source/developers/testing.rst [hunk 2]
```diff
@@ -484,7 +484,7 @@ Future Work on Unit Tests
 
 - Review pull requests from @bencomp for ideas for approaches to testing: https://github.com/IQSS/dataverse/pulls?q=is%3Apr+author%3Abencomp
 - Come up with a way to test commands: http://irclog.iq.harvard.edu/dataverse/2015-11-04#i_26750
-- Test EJBs using Arquillian, embedded app servers, or similar. @bmckinney kicked the tires on Arquillian at https://github.com/bmckinney/bio-dataverse/commit/2f243b1db1ca704a42cd0a5de329083763b7c37a
+- Test EJBs using Arquillian, embedded app servers, or similar. @bmckinney kicked the tires on Arquillian at https://github.com/bmckinney/bio-dataverse/commit/2f243b1db1ca704a42cd0a6de329083763b7c37a
```
[the remark this answers is not in the document]
Comment (pdurbin) [reply doc/sphinx-guides/source/developers/testing.rst:487]: Fixed in aedc4a3e42

File: src/main/java/edu/harvard/iq/dataverse/DatasetServiceBean.java [hunk 3]
```diff
@@ -514,7 +513,7 @@ public List<DatasetLock> listLocks(DatasetLock.Reason lockType, AuthenticatedUse
         }
         try {
             return query.getResultList();
-        } catch (javax.persistence.NoResultException e) {
+        } catch (jakarta.persistence.NoResultException e) {
```
[the remark this answers is not in the document]
Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/DatasetServiceBean.java:516]: Fixed in b689f67b93 along with many similar removals of jakarta noise.

File: src/main/java/edu/harvard/iq/dataverse/MailServiceBean.java [hunk 4]
```diff
@@ -124,7 +120,7 @@ public boolean sendSystemEmail(String to, String subject, String messageText, bo
                 try {
                     Transport.send(msg, recipients);
                     sent = true;
-                } catch (SMTPSendFailedException ssfe) {
+                } catch (MessagingException ssfe) {
                     logger.warning("Failed to send mail to: " + to);
                     logger.warning("SMTPSendFailedException Message: " + ssfe);
```
[the remark this answers is not in the document]
Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/MailServiceBean.java:125]: Updated in 9e4d837fc5

File: src/main/java/edu/harvard/iq/dataverse/harvest/client/HarvesterServiceBean.java [hunk 5]
```diff
@@ -148,12 +148,12 @@ public void doHarvest(DataverseRequest dataverseRequest, Long harvestingClientId
...
         String logTimestamp = logFormatter.format(new Date());
         Logger hdLogger = Logger.getLogger("edu.harvard.iq.dataverse.harvest.client.HarvesterServiceBean." + harvestingClientConfig.getName() + logTimestamp);
-        String logFileName = "../logs" + File.separator + "harvest_" + harvestingClientConfig.getName() + "_" + logTimestamp + ".log";
+        String logFileName = System.getProperty("com.sun.aas.instanceRoot") + File.separator + "logs" + File.separator + "harvest_" + harvestingClientConfig.getName() + "_" + logTimestamp + ".log";
         FileHandler fileHandler = new FileHandler(logFileName);
         hdLogger.setUseParentHandlers(false);
         hdLogger.addHandler(fileHandler);
         
-        PrintWriter importCleanupLog = new PrintWriter(new FileWriter( "../logs/harvest_cleanup_" + harvestingClientConfig.getName() + "_" + logTimestamp+".txt"));
+        PrintWriter importCleanupLog = new PrintWriter(new FileWriter(System.getProperty("com.sun.aas.instanceRoot") + File.separator + "logs/harvest_cleanup_" + harvestingClientConfig.getName() + "_" + logTimestamp + ".txt"));
```
[the remark this answers is not in the document]
Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/harvest/client/HarvesterServiceBean.java:156]: Looooong Slack discussion here: https://iqss.slack.com/archives/C010LA04BCG/p1690912135247139

**Comment by pdurbin** [comment]
I just moved this PR to "ready for review" because it is.

Here's what I'd like other reviewers to think about or notice:

- I'm in favor of removing the old Perl installer (and docker-aio and docker-dcm) but for now this change is in a different branch. We should make a decision about it: https://github.com/IQSS/dataverse/pull/9685#discussion_r1283596612
- I added a SwordFilter for backward compatibility: https://github.com/IQSS/dataverse/pull/9685#discussion_r1283641465
- Also for SWORD, bean discovery mode was changed: https://github.com/IQSS/dataverse/pull/9685#discussion_r1283651549
- Primepush was removed: https://github.com/IQSS/dataverse/pull/9685#discussion_r1283653212
- Dataset citation download was adjusted: https://github.com/IQSS/dataverse/pull/9685#discussion_r1283653989
- PrettyFaces StringUtils was replaced: https://github.com/IQSS/dataverse/pull/9685#discussion_r1283627142
- There is a fix for Docker related to finding domain1/logs: https://github.com/IQSS/dataverse/pull/9685#discussion_r1283646252

I left various other comments in my review that reviewers are welcome to check out. The above are the more important items, I feel.

**Review by rtreacy — COMMENTED** [review]
There are a lot of changes to many files here. The bulk of them are related to namespace changes from javax.* to jakarta.*, mostly in the import statements and there is also some cleanup of unused or redundant imports. Calls to Pretty Faces APIs for strings have been replaced. Some exception handling has been updated, for example in the mail service an old Sun SMTPSendFailedException has been  jakarta.mail.MessagingException. A new SwordFilter has been added that mutates headers to work with Payara 6
This all looks good to me, just want to see if anyone else has any feedback and want to be careful about getting in the way of polishing off 5.14

**Review by rtreacy — APPROVED** [review]

**Comment by kcondon** [comment]
Pls update Dataverse version to v5.14

**Comment by kcondon** [comment]
Issues found:
1. Repeated errors thrown in logs when doing nothing:
[idle_log_err.txt](https://github.com/IQSS/dataverse/files/12292023/idle_log_err.txt)

**Comment by poikilotherm** [comment]
> Issues found:
> [...]
>     3. Clicking on filename from tree view throws 500 error in ui, saw same context validation errs in log, same error just clicking on file in dataset file table.

Should be fixed with [3716c5c](https://github.com/IQSS/dataverse/pull/9685/commits/3716c5c4ffe9c2a02a3690e7ff6bba9d3a4590b9).

I scanned all XHTML code for other examples of some `<a>` with an `jsf:action` or `jsf:actionHandler` and could not find other appearances. Looks like dataset and file page citation widget were the only places.

**Comment by landreev** [comment]
Regarding item 1. above - those long "illegal State" and "validation failed" exception stack traces are very easy to reproduce: just try accessing a dataverse that does not exist in the database - http://localhost:8080/dataverse/blahblahblah
I'm seeing it in my own payara6 build environment too.

**Comment by poikilotherm** [comment]
> Regarding item 1. above - those long "illegal State" and "validation failed" exception stack traces are very easy to reproduce: just try accessing a dataverse that does not exist in the database.

I can verify this procedure and see error logs as well.

**Comment by scolapasta** [comment]
My vote for the kebab metadata issue above would be the 2nd fix: https://github.com/IQSS/dataverse/commit/fedb103d984d5bf67693f0034ae77855eb9bdce4
as I don't love referencing the DatasetPage on a multi-use bean. I'm also not too particularly concerned, as eventually all this will be going away.

**Comment by kcondon** [comment]
Merging this after discussing on tech. @scolapasta Re: P6 I'm ok with merge now and resolve remaining issues as top priorities.

**Pull request merged**